Add rate-limiter [#301](https://github.com/snoyberg/keter/issues/301)

Oleksandr-Zhabenko opened the pull request:
The following pull request adds keter rate limiter using specially created package keter-rate-limiting-plugin. It adds WAI middleware rate limiter with 5 possible algorithms that supports complex configurations. It addresses issue [#301 ](https://github.com/snoyberg/keter/issues/301)

jappeace reviewed (changes requested):
please find another way of encoding `middlewareCache`.
I don't think we should introduce global variables anywhere.

Oleksandr-Zhabenko commented on src/Keter/Proxy.hs:
```diff
@@ -73,7 +75,41 @@ import Paths_keter qualified as Pkg
+
+{-# NOINLINE middlewareCache #-}
+middlewareCache :: IORef (HM.HashMap MWCacheKey Wai.Middleware)
+middlewareCache = unsafePerformIO $ newIORef HM.empty
```
There is a possibility to do the following:

Migration checklist
Types:

* Change middleware fields in your stanza/config types to [MiddlewareSpec].
* Update FromJSON instances to parse via parseMiddlewareSpecs.
Wiring:

* Build a MiddlewareRegistry in Main (registerCoreBuiltins, registerRateLimiter, etc.).
* Pass the registry to Proxy.makeSettings.
Plugins:

* Move rate-limiter parsing and construction into its own module exporting registerRateLimiter.
* Any future middleware adds only a new registerX call in Main; no changes in Keter core.

This gives you “more a plugin”: Keter core no longer pattern matches on particular middlewares, and the rate limiter lives entirely as a plugin with a one-line registration.

Is this desirable?

jappeace commented on cabal.project.local~:
```diff
@@ -0,0 +1,3 @@
+ignore-project: False
+extra-include-dirs: /nix/store/cbdvjyn19y77m8l06n089x30v7irqz3j-zlib-1.3.1-dev/include
+extra-lib-dirs: /nix/store/l5g2v1jgfyf3j0jp9iv5b79fi8yrwzpp-zlib-1.3.1/lib
```
you probably don't want to commit this...

Oleksandr-Zhabenko commented on src/Keter/Config/Middleware.hs:
```diff
@@ -63,38 +85,62 @@ instance ToJSON MiddlewareConfig where
+toMiddlewareIO (BasicAuth realm cred) =
+  pure $ basicAuth (\u p -> pure $ (Just p ==) $ lookup u cred) (fromString realm)
+toMiddlewareIO (AddHeaders headers) = pure $ addHeaders headers
+toMiddlewareIO (RateLimiter rl)     = buildRateLimiter rl
```
I provided the requested changes:
* Env is constructed once (per compiled middleware chain) and reused; no per-request construction.
* All shared maps are guarded by TVar and accessed via STM; safe under green-threaded concurrency.
* Keter.Config.Middleware caches the composed middleware chain per host + config list (the TVar-based cache in Proxy), so each RateLimiter Env lives as long as that compiled chain, preserving counters and windows as intended.

Oleksandr-Zhabenko commented on src/Keter/Proxy.hs:
```diff
@@ -74,6 +78,40 @@ import Prelude hiding (FilePath, (++))
+-- TVar-backed cache, scoped via ProxySettings (no globals)
+newtype MiddlewareCache = MiddlewareCache
+  { mcVar :: TVar (HM.HashMap MWCacheKey Wai.Middleware)
+  }
```
> now we no longer need the cache right? because the middleware get's constructed by performAction, which creates a wai app, which then get's run. so performAction is only called when the app is created

@jappeace I am not sure I understand your comment. 

* 'MiddlewareCache' is exported from the module for possible usage mostly in the test suites.
* The data it provides is used now in withClient and performAction which are not exported outside the module. 
* Could we omit this new data type completely? Well, yes, but the data it represents are needed to be used to provide rate limiter functionality. 
* Would you like to omit creating separate data types and use the data themselves without natural declarations? Or would you like to move the data type definitions into the keter-rate-limiting-plugin codebase? Or do you state here that the cache has already been changed so that to provide more declarative way of definitions?

Oleksandr-Zhabenko replied:
I have implemented “compile-at-activation” middlewares and removed MiddlewareCache while preserving all functionality.

Notes:

* ProxyActionRaw now carries compiled Wai.Middleware instead of [MiddlewareConfig].
* App compiles middlewares once in withActions and no longer creates a per-app cache.
* HostManager stores compiled actions without a cache.
* Proxy applies the precompiled middleware directly.
* Common removes MiddlewareCache definitions.

Nevertheless, I removed automatic Show instance derivation for ProxyActionRaw and replaced it with manual one.

jezen commented on src/Keter/Common.hs:
```diff
@@ -3,6 +3,7 @@
 {-# LANGUAGE TypeFamilies #-}
 
 -- | Provides logging, versioning and some type aliases
+-- Updated: middlewares are compiled at activation.
```
I don't think this comment should be here.

jezen commented on src/Keter/Config/V10.hs:
```diff
@@ -4,6 +4,14 @@
+--     So webapp stanzas can declare edge middlewares (e.g., rate limiter)
+--     applied by Keter itself, without app code changes.
+--   - ProxyActionRaw now carries a compiled Wai.Middleware. 
+--     Middlewares are compiled at activation/reload time
```
Compiled in what sense? I'm guessing this is _compiled_ in the sense of composed. It's a little confusing using the word "compile" in the context of a language where the central part of our workflow is the Haskell compiler.

Oleksandr-Zhabenko replied:
I have changed it. Now it uses 'set up'.

jezen commented on src/Keter/Proxy.hs:
```diff
@@ -4,6 +4,7 @@
 {-# LANGUAGE TupleSections #-}
 
 -- | A light-weight, minimalistic reverse HTTP proxy.
+-- Updated: apply precompiled middlewares; remove MiddlewareCache.
```
This looks like ChatGPT output.

Oleksandr-Zhabenko commented:
> This is quite a large change, and there is clearly ChatGPT output everywhere which suggests to me that not enough care was taken in ensuring the changes here are actually the desired changes.
> 
> There's also no evidence that this solution actually works. In order to verify that, someone would have to clone the repository with this change, run it manually, and try to observe the expected rate limiting behaviour.
> 
> I would expect to see, for example, a NixOS test which runs a minimal example application within keter which is configured with a rate limiter, and is then attacked with [siege](https://linux.die.net/man/1/siege), or perhaps just a load of HTTP requests sent with cURL in a loop.

Actually, T found the needed requested changes by Jappie quite large and complicated and I used ChatGPT to provide them. Firstly, T needed to convince ChatGPT to make them (it suggested not to make such changes), then I provided the additional data for the AT to provide the changes safely Afterwards, I have carefully compared the changes with the oroginal versions using meld, and made mostly additions of the deleted original comments (but left 'Updated:' informatonal messages for easier review). Afterwards, I changed tests accordinglyfor the new APT. There was a need to add Show instance for ProxyActionRaw, using another A model. Then I just accurately tested build and test and then after rerun because of Mac0S random failure, provided the actual variant.

Actually, there are [two](https://github.com/Oleksandr-Zhabenko/keter-rate-limiting-plugin/blob/main/extended-integration-test.nix) nix [tests](https://github.com/Oleksandr-Zhabenko/keter-rate-limiting-plugin/blob/main/minimal-integration-test.nix) of the rate limiter functionality in the [keter-rate-litiming-plugin](https://hackage.haskell.org/package/keter-rate-limiting-plugin) package itself. They use legacy keter (without integration with the new package within). I will provide another one here, actually using integration layer, should I?

jezen commented:
> Actually, there are [two](https://github.com/Oleksandr-Zhabenko/keter-rate-limiting-plugin/blob/main/extended-integration-test.nix) nix [tests](https://github.com/Oleksandr-Zhabenko/keter-rate-limiting-plugin/blob/main/minimal-integration-test.nix) of the rate limiter functionality in the [keter-rate-litiming-plugin](https://hackage.haskell.org/package/keter-rate-limiting-plugin) package itself. They use legacy keter (without integration with the new package within). I will provide another one here, actually using integration layer, should I?

If I run keter, but this functionality isn't tested against keter, then I can't really have any confidence that this functionality works in keter.

Oleksandr-Zhabenko commented on etc/keter-config.yaml:
```diff
@@ -58,3 +58,41 @@ listeners:
 # cli-port: 1234
+#
+
+# Middlewares are configured per app in the bundle’s config/keter.yaml, not here.
+# Example (inside a bundle):
```
> I think an integrated test would be better, because it would both serve as an example that users can follow, and _also_ it will show evidence that the feature actually works.

'Interated' in what sense? Should it be in the incoming directory? Or in the root? It seems to be similar to the two [integration tests](https://github.com/Oleksandr-Zhabenko/keter-rate-limiting-plugin/blob/main/extended-integration-test.nix) in the keter-rate-limiting-plugin repository.

jezen replied:
In the sense that we can see a keter process running in a virtual machine, serving requests and exhibiting the rate limiting behaviour that this pull request is meant to achieve.

Oleksandr-Zhabenko replied:
I have also moved example config for middleware into incoming/middleware example.

Oleksandr-Zhabenko commented:
> does this work or not? @Oleksandr-Zhabenko I'm fine with merging as is.
> 
> also, the fact oleksander has to open a PR to add a plugin is just a major design flaw.
> 
> #319

Yes, @jappeace, it works.

Oleksandr-Zhabenko commented on ChangeLog.md:
```diff
@@ -1,5 +1,13 @@
+## 2.3.0
+
+- Add integration with keter-rate-limiting-plugin. [301](https://github.com/snoyberg/keter/issues/301)
+- Provided additional possiblities to apply middlewares per-app.
```
Fixed.

Oleksandr-Zhabenko commented on ChangeLog.md:
```diff
@@ -1,5 +1,13 @@
+- Provided additional possiblities to apply middlewares per-app.
+- Added some integration tests for the new functionality.
+- Change some function signatures in dependent modules for integration.
+- Some documentation improvements.
```
Done as requested.

Oleksandr-Zhabenko commented on incoming/middleware/config/keter.yaml:
In the incoming there are other specific examples, therefore, I have decided earlier to provide example configuration there.

jappeace commented on minimal-rate-limiter-integration.nix:
```diff
@@ -0,0 +1,243 @@
+app :: Application
+app _ respond =
+  respond $ responseLBS status200 [("Content-Type","text/plain")] "ok from rate-limited backend"
+HS
```
this is bonkers 😅 , can't you like add a hs file to the repo to copy in the build for a test?

Oleksandr-Zhabenko replied:
I have moved heredoc to a separate test/Backend.hs file.

Oleksandr-Zhabenko commented on minimal-rate-limiter-integration.nix:
```diff
@@ -0,0 +1,243 @@
+    print("=== Verify rate limiting is active ===")
+    status, series = machine.execute("bash -lc 'for i in $(seq 1 20); do curl -s -o /dev/null -w \"%{http_code}\\n\" -H \"Host: rl.test\" http://127.0.0.1:${toString port}/; sleep 0.1; done'")
+    print("Burst test responses: {0}".format(series))
+    assert '429' in series, "Expected 429 responses in burst test, got: {0}".format(series)
```
> does the main branch fail if you run this test against it?
No, it succeeds. It works.

Oleksandr-Zhabenko commented:
@jappeace @jezen @ulidtko I have updated the keter-rate-limiting-plugin version to 0.2.0.2, fixed IO issue with integration tests .nix files. 

Now it is also [available](https://www.stackage.org/nightly-2025-10-08/package/keter-rate-limiting-plugin-0.2.0.2) on Stackage (nightly builds). 

Besides I have created a [separate dedicated repository](https://github.com/Oleksandr-Zhabenko/keter-rate-limiting-plugin-testing-output-files) with 3 zipped textual output files that contain the output of my .nix integration tests while being run (I truncated the OS introductory loading output and preserved the actual tests output). They are evidence of the passed tests (2 for the keter-rate-limiting-plugin itself and 1 for the keter integration).

Could you review the pull request here and merge it, please?

ulidtko commented on etc/keter-config.yaml:
```diff
@@ -58,3 +58,5 @@ listeners:
 
 # set a debug port
 # cli-port: 1234
+#
```
+1, it may feel a bit weird at first — but try to imagine yourself in like 8 months later, getting back to this codebase; forgot every detail and starts cursing WTF at the past self all over the place.

Please try to prevent that, as much as possible

jappeace commented on keter.cabal:
```diff
@@ -47,6 +47,7 @@ library
     , http-reverse-proxy    >=0.6.2    && <0.7
     , http-types            >=0.12.3   && <0.13
     , indexed-traversable   >=0.1.2    && <0.2
+    , keter-rate-limiting-plugin ==0.2.0.2
```
wouldn't < 0.2.1 be better so you get minor updates automatically?
